### PR TITLE
Always Off Sampler

### PR DESCRIPTION
**Pull request by nholbrook**
Implements AlwayOffSampler based on the sampler interface in [#118](https://github.com/open-telemetry/opentelemetry-cpp/pull/118).

### CODE REVIEW

**Comment by nholbrook** [comment]
I signed it

File: sdk/include/opentelemetry/sdk/trace/sampler.h [hunk 1]
```diff
@@ -0,0 +1,90 @@
+#pragma once
```
Comment (jsandler18) [inline sdk/include/opentelemetry/sdk/trace/sampler.h:1]: This file is just copied from Oliver's PR, right?

File: sdk/src/trace/always_off_sampler.cc [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    const SpanContext *parent_context,
+    trace_api::TraceId trace_id,
+    nostd::string_view name,
+    trace_api::SpanKind span_kind,
+    const nostd::span<AttributeKeyValue> &attributes) noexcept
+  {
+    return
+    {
+      Decision::NOT_RECORD,
+      std::unique_ptr<nostd::span<AttributeKeyValue>>(new nostd::span<AttributeKeyValue>())
```
Comment (pyohannes) [inline sdk/src/trace/always_off_sampler.cc:20]: You can make this a `{ Decisions::NOT_RECORD, nullptr }`. I think we should agree on returning a `nullptr` in cases when no attributes are added, as this saves heap spaces and gets rid of an obsolete allocation (for every span).

File: sdk/include/opentelemetry/sdk/trace/sampler.h (hunk 1 above)
Comment (nholbrook) [reply sdk/include/opentelemetry/sdk/trace/sampler.h:1]: Yep

File: sdk/src/trace/always_off_sampler.cc (hunk 2 above)
Comment (nholbrook) [reply sdk/src/trace/always_off_sampler.cc]: Sounds good

File: sdk/src/trace/always_off_sampler.cc [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+namespace trace_api = opentelemetry::trace;
+
+SamplingResult AlwaysOffSampler::ShouldSample(
+    const SpanContext *parent_context,
+    trace_api::TraceId trace_id,
+    nostd::string_view name,
+    trace_api::SpanKind span_kind,
+    const nostd::span<AttributeKeyValue> &attributes) noexcept
+  {
+    return { Decision::NOT_RECORD, nullptr };
```
Comment (g-easy) [inline sdk/src/trace/always_off_sampler.cc:17]: Given the straightforward implementation, should this entire file be inlined into the header, in order to have fewer files overall and possibly benefit from inlining?

Comment (nholbrook) [reply sdk/src/trace/always_off_sampler.cc]: Yeah, I talked with @ziqizh and we both think that makes sense. I've moved the functionality to the header.

File: sdk/include/opentelemetry/sdk/trace/always_off_sampler.h [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+   * @return Description MUST be AlwaysOffSampler
+   */
+  std::string GetDescription() const noexcept override
+  {
+    return "AlwaysOffSampler";
+  }
+};
+}  // namespace trace
+}  // namespace sdk
+OPENTELEMETRY_END_NAMESPACE
```
Comment (reyang) [inline sdk/include/opentelemetry/sdk/trace/always_off_sampler.h:41]: nit: add blank line before EOF to make GitHub happy.

File: sdk/include/opentelemetry/sdk/trace/always_off_sampler.h [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+ */
+class AlwaysOffSampler : public Sampler
+{
+public:
+  /**
+   * @return Returns NOT_RECORD always
+   */
+  SamplingResult ShouldSample(
+    const SpanContext *parent_context,
+    trace_api::TraceId trace_id,
```
Comment (reyang) [inline sdk/include/opentelemetry/sdk/trace/always_off_sampler.h:23]: There is potential optimization to this, by avoiding the trace_id creation (random number generation) at all. Probably worth creating an issue for tracking.

Comment (nholbrook) [reply sdk/include/opentelemetry/sdk/trace/always_off_sampler.h:23]: Alright, I'll bring this up to Johannes later today.

File: sdk/include/opentelemetry/sdk/trace/sampler.h [hunk 6]
```diff
@@ -38,7 +38,7 @@ struct SamplingResult
 {
   Decision decision;
   // A set of span Attributes that will also be added to the Span. Can be nullptr.
-  std::unique_ptr<std::map<std::string, openelemetry::common::AttributeValue>> attributes;
+  std::unique_ptr<const std::map<std::string, opentelemetry::common::AttributeValue>> attributes;
```
Comment (pyohannes) [inline sdk/include/opentelemetry/sdk/trace/sampler.h:41]: Please rebase to get those changes in.

File: sdk/include/opentelemetry/sdk/trace/always_off_sampler.h [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    nostd::string_view name,
+    trace_api::SpanKind span_kind,
+    const trace_api::KeyValueIterable &attributes) noexcept override
```
Comment (pyohannes) [inline sdk/include/opentelemetry/sdk/trace/always_off_sampler.h:26]: We try to be consistent in [not naming unused parameters](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#f9-unused-parameters-should-be-unnamed). See [here](https://github.com/open-telemetry/opentelemetry-cpp/blob/master/api/include/opentelemetry/trace/noop.h#L38) for an example.

File: sdk/include/opentelemetry/sdk/trace/always_off_sampler.h [hunk 8]
```diff
@@ -0,0 +1,41 @@
+#pragma once
```
Comment (pyohannes) [inline sdk/include/opentelemetry/sdk/trace/always_off_sampler.h:1]: Let's sync with #122 to determine a proper location for those samplers. #122 puts the sampler implementation in `sampler.h`.

I prefer the interface definition to be in a separate file and not mixed with implementation, so the solution in this PR is preferable. Maybe we should have a `sdk/trace/samplers/always_off.h`?

File: sdk/include/opentelemetry/sdk/trace/always_off_sampler.h (hunk 5 above)
Comment (pyohannes) [reply sdk/include/opentelemetry/sdk/trace/always_off_sampler.h:23]: I don't see how we could easily avoid that. The `trace_id` is created only _once_ for every trace anyway. And even if an always-off sampler is active, we might still need the trace id for an outgoing context.

Also, the spec requires the trace id to be passed to a sampler.

File: sdk/include/opentelemetry/sdk/trace/always_off_sampler.h (hunk 8 above)
Comment (nholbrook) [reply sdk/include/opentelemetry/sdk/trace/always_off_sampler.h:1]: I like the idea of having the samplers in their own directory like you suggested. Let's touch base with @ziqizh during our sync and get his thoughts.

Comment (ziqizh) [reply sdk/include/opentelemetry/sdk/trace/always_off_sampler.h:1]: Agreed. I thought we wanted to put everything in `sampler.h`, but making a separate folder looks good to me. I will move the AlwaysOnSampler to `sdk/trace/samplers/always_on.h`

**Comment by nholbrook** [comment]
I've made all the requested changes and moved the sampler to `sdk/trace/samplers/always_off.h` following conversation with @pyohannes.

**Pull request merged**